arkouda/timeclass.py

```diff
@@ -144,6 +146,155 @@ def __init__(self, array, unit: str = _BASE_UNIT):  # type: ignore
+            self._microsecond = cast((self.values / _get_factor("us")), int64) % 1000
+            self._nanosecond = cast((self.values / _get_factor("us")), int64) % 1000
+
+            # Month, day
+            self._day = (self.values // _get_factor("d") + 1401) % 1461
+            self._day[self._day >= 731] -= 731
+            self._day[self._day >= 365] -= 365
+
+            self._month = self._day - self._day + 3
+            ge_306 = self._day >= 306
+            self._month[ge_306] = 1
+            self._day[ge_306] -= 306
+
+            ge_153 = self._day >= 153
+            self._month[ge_153] = 8
+            self._day[ge_153] -= 153
+
+            ge_122 = self._day >= 122
+            self._month[ge_122] += 4
+            self._day[ge_122] -= 122
+
+            ge_61 = self._day >= 61
+            self._month[ge_61] += 2
+            self._day[ge_61] -= 61
+
+            self._day += 2
+            ge_32 = self._day >= 32
+            self._month[ge_32] += 1
+            self._month[self._month >= 13] = 1
+            self._day[ge_32] -= 31
+
+            # Day of year
+            month_days = [
+                [0, 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334],
+                [0, 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335],
+            ]
+
+            days = self._day[:]
+            self._date = []
+            for i in range(days.size):
+                if self._month[i] == 2 and self._day[i] == 29 and _isLeapYear(self._year[i]) == 0:
+                    self._month[i] += 1
+                    self._day[i] = 1
+                    days[i] += month_days[_isLeapYear(self._year[i])][self._month[i] - 1]
+                else:
+                    days[i] += month_days[_isLeapYear(self._year[i])][self._month[i]]
+                self._date.append(f"{self._year[i]}-{self._month[i]:02d}-{self._day[i]:02d}")
+            self._dayofyear = array(days, int64)
+
+            # Day of week
+            self._dayofweek = cast(((self.values / _get_factor("d")) + 3), int64)
+            self._dayofweek %= 7
+
+            # Week of year
+            week = (self._dayofyear[:] - 1) - self._dayofweek + 3
+            week[week >= 0] //= 7
+            week[week >= 0] += 1
+
+            for i in range(days.size):
+                if week[i] < 0:
+                    year = int(self._year[i] - 1)
+                    if week[i] > -2 or (week[i] == -2 and _isLeapYear(year)):
+                        week[i] = 53
+                    else:
+                        week[i] = 52
+                elif week[i] == 53:
+                    if 31 - self._day[i] + self.dayofweek[i] < 3:
+                        week[i] = 1
+            self._weekofyear = week
+
+    # Datetime values
+    @property
+    def date(self):
+        return self._date
+
+    @property
+    def year(self):
+        return self._year
+
+    @property
+    def month(self):
+        return self._month
```

Comment:
we need to move the code for calculating these attributes into the property functions to avoid the unnecessary overhead like reuster said in his review. Something like the follow would work:

```python
def __init__(self, pda, unit: str = _BASE_UNIT):
  ...
self._date = None
self._year = None
  ...

@property
def date(self):
  if self._date is None:
    self._date = # do code to calculate date
  return self._date

@property
def year(self):
  if self._date is None:
    self._year = cast((days // 365.24 + 1970), int64)
  return  self._year
```
This way we only do the computation when a user asks for that attribute. But we save our work, so we don't have to recompute if they ask for it again